apm-agent-core/src/main/java/co/elastic/apm/agent/util/InstrumentationUsageUtil.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.util;
+
+import co.elastic.apm.agent.sdk.ElasticApmInstrumentation;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public final class InstrumentationUsageUtil {
+
+    private static final Set<ElasticApmInstrumentation> allInstrumentations = new HashSet<>();
+
+    private static final ConcurrentMap<ElasticApmInstrumentation, Boolean> usedInstrumentations = new ConcurrentHashMap<>();
```

Comment:
If there's no reason why we can't, I think we should make all of the fields and methods non-static and rename to something like `InstrumentationUsageReporter/Logger`.
I think static state should be confined to the minimum set of classes that really need to be static. While we can't really get rid of static state in `ElasticApmAgent`, I think we can do here. `ElasticApmAgent` would then have a static field of this class.

Reply:
Yeah, a lifecycle listener makes sense. I guess the static state would be the easiest then. But instead of including it in `MatcherTimerLifecycleListener`, I'd make `InstrumentationUsageUtil` implement `LifecycleListener`. In `stop`, it would then log the static state.
Ideally though, we would introduce more lifecycle events, such as `onInstrumentationInitialized` and `onInstrumentrationMatched`. But that might be over-engineered for that one use-case. If we see the need for these lifecycle methods more often, we can still do it then.

Reply:
Ha, race condition with c518c74 and b773b4b. That works for me as well.

Reply:
Should is still move the logging? Is that ok for you @eyalkoren?

Reply:
Looks [b729244](https://github.com/elastic/apm-agent-java/pull/2448/commits/b72924488637248a21396ca4d61300782314d505) good?